hoodie-client/src/main/java/com/uber/hoodie/table/DefaultBulkInsertPartitioner.java

```diff
@@ -0,0 +1,187 @@
+ * - Output spark partition will have records from only one hoodie partition.
+ */
+public class DefaultBulkInsertPartitioner<T extends HoodieRecordPayload> extends Partitioner implements BulkInsertPartitioner<T> {
+    private static Logger logger = LogManager.getLogger(DefaultBulkInsertPartitioner.class);
+    private static final int BUCKET_MULTIPLIER = 4;
+    private int expectedOutputSparkPartitions;
+    private int numBuckets;
+    private HoodiePartitionMapper hoodiePartitionMapper;
+    private int numOfPartitions = -1;
+
+    private void init(JavaRDD<HoodieRecord<T>> records, int outputSparkPartitions) {
+        expectedOutputSparkPartitions = Math.max(1, outputSparkPartitions);
+        numBuckets = expectedOutputSparkPartitions * BUCKET_MULTIPLIER;
+        this.hoodiePartitionMapper = new HoodiePartitionMapper();
+        TreeMap<String, Long> bucketRecordCount = mapRecordsToBucket(records);
+        Counter totalRecordCount = new Counter(0);
+        // Count total records.
+        bucketRecordCount.values().forEach(v -> totalRecordCount.addCount(v));
+        long expectedRecordsPerSparkPartition = totalRecordCount.getCount() / expectedOutputSparkPartitions;
+        Map<String, Counter> currHoodiePartitionRecordCnt = new HashMap<>();
+        for (Entry<String, Long> entry : bucketRecordCount.entrySet()) {
+            String[] parts = entry.getKey().split(" ");
+            String hoodiePartition = parts[0];
+            if (!currHoodiePartitionRecordCnt.containsKey(hoodiePartition)) {
+                currHoodiePartitionRecordCnt.put(hoodiePartition, new Counter(0));
+                numOfPartitions++;
+            }
+            Counter cnt = currHoodiePartitionRecordCnt.get(hoodiePartition);
+            if (cnt.getCount() >= expectedRecordsPerSparkPartition) {
+                // Reset count and also change output spark partition. That way going forward records for same hoodie
+                // partition will be sent to different spark output partition.
+                cnt.setCount(0);
+                numOfPartitions++;
+            } else {
+                cnt.addCount(entry.getValue());
+            }
+            hoodiePartitionMapper.add(entry.getKey(), numOfPartitions);
+        }
+    }
+
+    @Override
+    public int getPartition(final Object key) {
+        Tuple2<HoodieRecord<T>, Integer> pair = (Tuple2<HoodieRecord<T>, Integer>) key;
+        return hoodiePartitionMapper.get(getBucketKey(pair._1(), pair._2()));
+    }
+
+    @Override
+    public int numPartitions() {
+        return numOfPartitions + 1;
+    }
+
+    public JavaRDD<HoodieRecord<T>> repartitionRecords(JavaRDD<HoodieRecord<T>> records, int outputSparkPartitions) {
+        init(records, outputSparkPartitions);
+        // As we are going over input records twice; we need to cache input records.
+        JavaRDD<HoodieRecord<T>> cachedRecords = records.persist(StorageLevel.MEMORY_AND_DISK_SER());
```

Comment:
The reason why we are doing caching (DISK+MEMORY) is because we are reading it twice. So it is good if we cache it even for large datasets.

For SortBy - I do see that initial boundary computation may not need spark to read complete data but to later sort them it will have to read the entire data (per partition/executor) right? am I missing anything here?

I have collected the results for 64GB dataset with 32 executors and 16 hoodie writers (insertParallelism). [This is almost equivalent to 1TB for 512 executors]. 

With SortBy
<img width="1426" alt="screen shot 2017-08-20 at 12 29 22 pm" src="https://user-images.githubusercontent.com/899843/29497784-4f4660b6-85a3-11e7-9b78-34bdb00ea02e.png">

With Custom partitioner

<img width="1426" alt="screen shot 2017-08-20 at 12 29 57 pm" src="https://user-images.githubusercontent.com/899843/29497787-57544692-85a3-11e7-88d7-1b2427076ece.png">